lightning/src/util/ser.rs

```diff
@@ -34,65 +34,45 @@ use util::byte_utils::{be64_to_array, be48_to_array, be32_to_array, be16_to_arra
 /// serialization buffer size
 pub const MAX_BUF_SIZE: usize = 64 * 1024;
 
-/// A trait that is similar to std::io::Write but has one extra function which can be used to size
-/// buffers being written into.
-/// An impl is provided for any type that also impls std::io::Write which simply ignores size
-/// hints.
+/// A struct that implements a serialization buffer.  A size hint can be given and the length
+/// of the result can be retrieved.
 ///
-/// (C-not exported) as we only export serialization to/from byte arrays instead
-pub trait Writer {
+/// std::io::Write is implemented for serializing Bitcoin objects with impl_consensus_ser! below.
+pub struct Writer(pub Vec<u8>);
+
+impl Writer {
+	/// Construct an empty Writer
+	pub fn new() -> Self {
+		Writer(Vec::new())
+	}
+
 	/// Writes the given buf out. See std::io::Write::write_all for more
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error>;
+	pub fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+
 	/// Hints that data of the given size is about the be written. This may not always be called
 	/// prior to data being written and may be safely ignored.
-	fn size_hint(&mut self, size: usize);
-}
+	pub fn size_hint(&mut self, size: usize) {
+		self.0.reserve(size)
+	}
```

Comment:
I was looking into how `reserve` is implemented. It seems it increases the capacity to the _max_ of either the desired capacity or double the current capacity. This makes me wonder how useful `size_hint` really is. 

I suppose we may save some allocations if the writer has 0 capacity, but I'm wondering if the same thing can be accomplished while defining our public interfaces in terms of `std::io::Write` instead of `Writer`. For instance, if a message knows it size, we could do something like:

```rust
Writer::with_capacity(message.size_hint());
```

Rather than having the message call `size_hint` on `Writer`. Thoughts?

Reply:
We could do `reserve_exact`.

But I am guessing that the caller almost always calls these with a zero sized writer, otherwise it's kind of ineffective.  If this is called with a non-zero sized writer (e.g. if this object is a field in a larger struct), then, yes, the size_hint mechanism won't be very effective and we should think about moving to something like `with_capacity`.

Maybe @TheBlueMatt has more insight about how `size_hint` is used.

Reply:
Hmm, in  general it probably doesn't make sense to use it for sub-structs. It's mostly intended for where we know the exact size, which we generally do in all our P2P message sending, avoiding copying the data a few times. Luckily, the max Jeff referenced means we can't really do worse than the "normal" behavior with it, though.

Reply:
As for putting it in messages vs calling it on a Writer, I certainly don't care which we end to with, but it should only be one or the other, and this PR is already nontrivial in size.

Reply:
> Hmm, in general it probably doesn't make sense to use it for sub-structs.

It seems to be getting called for sub-structs, given that it's called on non-empty `Writer` instances during the tests.  Should I open an issue to look into that?

Reply:
> It seems to be getting called for sub-structs, given that it's called on non-empty Writer instances during the tests. Should I open an issue to look into that?

I mean, I guess its not strictly a bug, though it would be nice to create a sensible estimate at the top of any write methods and call it there, then we could drop it for most sub-structs.

Reply:
My main gripe is that we are restricting our interface since any `std::io::Write` can no longer be used, though we can always re-expand later, I suppose.

For size hints, the nice properties of the message returning the size are:

(1) We can use `std::io::Write` in our interfaces without needing another trait.
(2) Size calculations can be encapsulated in sub-structs rather than enclosing structs needing to know how to calculate the sizes of their parts.

Reply:
> My main gripe is that we are restricting our interface since any std::io::Write can no longer be used, though we can always re-expand later, I suppose.

I'm a little confused, is your question about not wanting to do the concrete-ization in this PR at all, or just that the sizing API isn't great? I think I agree with you on the sizing API, we should change it, but this PR doesn't touch it, so I figure that can come in a followup.

Reply:
Yeah, the former. The sizing API change can come later. I had thought they may be related if any method taking a `Writer` was also using `size_hint`. But maybe only direct instantiators of `Writer` (`VecWriter` before) use it?

So in this PR, I had figured `Writer` would implement `std::io::Write` so the interface changes would therefore be limited to changing traits and not break any potential users. But maybe there is something related to generifying `Sign` that I'm missing. It's not immediately clear to me how/if these two things are connected. Could use some clarification in the PR description if so.

Reply:
I updated the PR description.

The relevant method in `Writeable` is `write(&self, writer: &mut Writer)`.  If we want to use a trait (e.g. `Write`) instead of the concrete `Writer`, that would have to change to either:

- `write<W: Write>(&self, writer: &mut W)` - this is basically back to the previous code, which would cause `Sign` to have a generic method, and therefore not be able to be `dyn`
- `write(&self, writer: &mut dyn Write)` - this is possible, with a slight performance impact

Reply:
Ah, is this because `ChannelMonitor` needs to be `Writeable` and thus so does `Sign`? Couldn't we parameterize as `ChannelMonitor<Signer: Sign + Writeable>`?

Reply:
> Couldn't we parameterize as ChannelMonitor<Signer: Sign + Writeable>?

Yes, but ultimately, as far as I understand, that doesn't solve the issue - as soon as you `Box<dyn Trait>` you can't have generics on the individual method definitions. In theory we could do `Sign<Writer>`, but of course then you're stuck with one writer at compile-time and can't have multiple. It may be possible to do that and then pick `dyn Writer` at compile time, but that would drop a *ton* of optimization we're basically relying on for users who want multiple-different-writer support.

Another approach which doesn't "solve" the problem, but may help work around it, is we could move the `write` on `Sign` to `KeysManager` (ie have a `KeysManager::write_sign()` method), which would allow for a `dyn Sign`, but not a `dyn KeysManager`.

Reply:
> Yes, but ultimately, as far as I understand, that doesn't solve the issue - as soon as you Box<dyn Trait> you can't have generics on the individual method definitions

You could do `Box<dyn Sign>` but not `Box<dyn Sign+Writeable>`.  So that would still work for having multiple `Sign` implementations at runtime (e.g. dispatched via a `DynSigner` as discussed elsewhere), but they wouldn't be implementing their own serialization (`Writeable`).

Reply:
> but they wouldn't be implementing their own serialization (Writeable).

Right, so we'd have to do my second suggestion, above, and move the serialization to the KeysManager in some way?

Reply:
Right, presumably via KeysManager or some `Sign`-specific write API that does what this PR does but only for `Sign`. Thoughts, @jkczyz ?

Reply:
@TheBlueMatt Not sure I understand how that would look with regards to a `ChannelMonitor` serializing its parameterized `Sign`.

@devrandom Could `DynSigner` also have a field storing something like `FnOnce` that downcasts the boxed signer and calls write. Then `DynSigner` could implement `Writeable` by calling this. The instantiator of `DynSigner` would need to give this function, but presumably that is possible since they are also giving boxed signer and thus know what type to downcast to.

Reply:
> @TheBlueMatt Not sure I understand how that would look with regards to a ChannelMonitor serializing its parameterized Sign.

Right, I guess there's a few options - take a reference to the `KeysManager` and call `KeysManager::serialize_signer(sign: &Sign)`. Or we could go with a straightforward `Sign::serialize() -> Vec<u8>` only.

Reply:
> @devrandom Could `DynSigner` also have a field storing something like `FnOnce` that downcasts the boxed signer and calls write? Then `DynSigner` could implement `Writeable` by calling this. The instantiator of `DynSigner` would need to give this function, but presumably that is possible since they are also giving boxed signer and thus know what type to downcast to.

Yes, that works.  There are actually more options, since a generic `Write` can be converted to a `dyn Write` inside `DynSigner`.

But in any case, moving `Writeable` from `BaseSign` to `Sign` is definitely a valid approach.  I am not worried about the sunk cost of having rewritten the `Writeable` trait, happy to back that out.  Should decide whether or not to do that based on whether we actually expect some advantage from not going via a Vec intermediary serialization.

Reply:
Here's a branch with the above implemented: https://github.com/rust-bitcoin/rust-lightning/compare/main...lightning-signer:degenerify1?expand=1

Reply:
> Yes, that works.

Huh! I guess when the limits the language is applying are arbitrary, the workarounds are arbitrary too :).

That looks good to me, too, what do you think @jkczyz?

Reply:
> Huh! I guess when the limits the language is applying are arbitrary, the workarounds are arbitrary too :).
> 
> That looks good to me, too, what do you think @jkczyz?

Works for me! Would be nice to move `Writeable` off to where it is needed rather than making it a sub-trait, if that's possible. Wasn't quite clear if the workaround allowed for that. But either way seems clean enough.

> I am not worried about the sunk cost of having rewritten the `Writeable` trait, happy to back that out.

@devrandom Regarding `Writer` trait, I do like making that a struct and using `std::io::Write`, but that is no longer applicable to this PR and can come at some later time, as you said. I think you meant `Writer` here, at least.